feat(datasets): add Shopify product-catalogue VLM dataset (#3400)
NVIDIA-NeMo/Automodel: docs/guides/dataset-overview.mdx

Reviewer: ```suggestion
- If a dataset you need is missing, open a [GitHub issue](https://github.com/NVIDIA-NeMo/Automodel/issues) with a short description and an example schema so the team can prioritize support.

```

Reviewer: ```suggestion
NeMo AutoModel supports several common patterns for language modeling and instruction tuning.

```

Reviewer: ```suggestion
- Example YAML:

```

Reviewer: ```suggestion
- Use case: instruction or QA tuning with either prompt-and-answer formatting or chat-template formatting

```

Reviewer: ```suggestion
- You can use the optional `seq_length` parameter for padding or truncation.
```

Reviewer: ```suggestion
</Note>

```

Reviewer: ```suggestion
- Example YAML:

```

Reviewer: ```suggestion
### ColumnMappedTextInstructionDataset (Generic Instruction SFT)
- Class: `nemo_automodel.components.datasets.llm.column_mapped_text_instruction_dataset.ColumnMappedTextInstructionDataset`
- Use case: quickly adapt instruction datasets by mapping your schema columns to `context`, `question`, and `answer`
- Sources: local JSON/JSONL or Hugging Face Hub dataset ID
- Notes:
  - Set `use_hf_chat_template: true` to opt into a usable tokenizer chat template. With `answer_only_loss_mask: true`, the dataset masks non-answer tokens. Without a usable template, it falls back to prompt-and-answer formatting.
  - Map-style, non-streaming dataset (supports `len(ds)` and `ds[i]`)
  - For streaming, including Delta Lake or Databricks, use `ColumnMappedTextInstructionIterableDataset` (see [Streaming Datasets](#streaming-datasets)).
- Example YAML:

```

Reviewer: ```suggestion

See the detailed guide, [Column-Mapped Text Instruction Dataset](/datasets/columnmapped-dataset), for more information.
```

Reviewer: ```suggestion
  - `tokenizer`: tokenizer instance (required and must have chat template support)
  - `split`: dataset split (for example, "train" or "validation")
  - `name`: dataset configuration or subset name
  - `seq_length`: maximum sequence length for padding or truncation
  - `padding`: padding strategy (for example, "do_not_pad" or "max_length")
  - `truncation`: truncation strategy (for example, "do_not_truncate" or "longest_first")
  - `chat_template`: optional override for the tokenizer chat template
```

Reviewer: ```suggestion
  - `skip_invalid_samples`: if `true`, skip malformed JSONL lines when reading local files (warnings log skip counts). The default `false` fails fast on a bad line

```

Reviewer: ```suggestion
- Assistant messages might also include `reasoning_content` for structured reasoning traces
```

Reviewer: ```suggestion
</Note>

```

Reviewer: ```suggestion
- Example YAML:

```

Reviewer: ```suggestion

The following JSON shows the expected OpenAI messages format:

```

Reviewer: ```suggestion

The following example shows the template requirement for `reasoning_content`:

```

Reviewer: ```suggestion

For single-turn tool calling (one tool call per conversation), omit the tool response and final assistant message:

```

Reviewer: ```suggestion

See the [Function Calling guide](/recipes-e2e-examples/function-calling) for an end-to-end example with FunctionGemma.
```

Reviewer: ```suggestion

Refer to [Retrieval dataset](/datasets/retrieval-dataset) for more information. For the end-to-end
```

Reviewer: ```suggestion
- Use case: token-level LM pretraining over `.bin` shards produced by NanoGPT-style preprocessors (supports legacy and current formats)

```

Reviewer: ```suggestion
### Megatron Pretraining and Interoperability with Pre-Tokenized Megatron Data
```

Reviewer: ```suggestion
- Interoperability: If your corpus has already been tokenized and indexed for Megatron (that is, `.bin` and `.idx` pairs), you can point AutoModel to those assets directly. No re-tokenization is required.
```

Reviewer: ```suggestion
- Example YAML:

```

Reviewer: ```suggestion

See the detailed [pretraining guide](/recipes-e2e-examples/pretraining), which uses MegatronPretraining data.
```

Reviewer: ```suggestion
Streaming datasets load and process data incrementally, one batch at a time, rather than loading the entire dataset into memory upfront. This approach provides the following benefits:
```

Reviewer: ```suggestion
- Available memory is limited compared to the dataset size
```

Reviewer: ```suggestion
- You are experimenting with a subset of a large dataset
```

Reviewer: ```suggestion
- You need random access to samples, such as for certain sampling strategies
```

Reviewer: ```suggestion
- Streaming reduces memory usage but might be slower than in-memory datasets
- Network latency can affect streaming performance for remote datasets
```

Reviewer: ```suggestion
- A buffer size of 10,000–100,000 samples typically provides a good balance
- For perfect shuffling, you need a buffer size equal to the dataset size, which defeats the purpose of streaming
```

Reviewer: ```suggestion
The following YAML shows a complete streaming example for a large instruction-tuning dataset:
```

Reviewer: ```suggestion
This configuration provides the following behavior:

```

Reviewer: ```suggestion
To reduce padding and improve throughput with variable-length sequences:

```

Reviewer: ```suggestion

The fine-tuning recipe passes these settings to its packed-dataset path. For THD packing, it disables packing when the model does not support `seq_lens`.
```

Reviewer: ```suggestion
Use a collator that pads to an FP8-friendly multiple when training with FP8:

```

Reviewer: ```suggestion
## VLM Datasets (Vision, Audio, and Language)
```

Reviewer: ```suggestion
Select a collate function in your YAML config:

```

Reviewer: ```suggestion

Answer-only loss masking is built into the supported VLM collators. They derive assistant regions from the processor's chat template with `build_labels_from_template`. The `start_of_response_token` parameter is not a collate argument.
```

Reviewer: ```suggestion
Diffusion models do not train directly on raw images or videos. Instead, the data is first encoded into a compact numerical representation called a latent, which is what the model learns from. Text captions are similarly converted into text embeddings that the model uses as conditioning.
```

Reviewer: ```suggestion
This encoding occurs during preprocessing, and the results are saved as `.meta` cache files. Training then reads these cache files directly, which is significantly faster than re-encoding on every step.
```

Reviewer: ```suggestion
The built-in preprocessing tool ([`tools/diffusion/preprocessing_multiprocess.py`](https://github.com/NVIDIA-NeMo/Automodel/blob/main/tools/diffusion/preprocessing_multiprocess.py)) handles this conversion. It uses a variational autoencoder (VAE) to encode visual data and a text encoder for captions, grouping outputs into resolution-bucketed directories compatible with the multiresolution data loader.
```

Reviewer: ```suggestion
- **Video (T2V)**: `nemo_automodel.components.datasets.diffusion.build_video_multiresolution_dataloader` for Wan 2.1 and HunyuanVideo
- **Image (T2I)**: `nemo_automodel.components.datasets.diffusion.build_text_to_image_multiresolution_dataloader` for FLUX.1-dev
```

Reviewer: ```suggestion
You can integrate custom datasets with zero code changes to NeMo AutoModel by using `_target_` in YAML. Use one of the following approaches.
```

Reviewer: ```suggestion
- LLM example (class):

```

Reviewer: ```suggestion

- LLM example (factory function):

```

Reviewer: ```suggestion

- VLM example (factory function):

```

Reviewer: ```suggestion
### Point to a Local Python File and Function

```

Reviewer: ```suggestion

The `build_my_dataset` function returns either a `datasets.Dataset` or a list or iterator of conversation dictionaries for VLM.
```

Reviewer: ```suggestion
- Ideal when your data has columns such as `instruction`, `input`, or `output` but with arbitrary names
- Supports local JSON/JSONL and Hugging Face Hub

```

Reviewer: ```suggestion
If you prefer Python, implement `get_context` and `get_target` and reuse the built-in preprocessor:

```

Reviewer: ```suggestion

Then reference your class with `_target_` in YAML.
```

Reviewer: ```suggestion
- **Chat templates**: SQuAD uses a tokenizer chat template automatically when available. Column-mapped datasets opt in with `use_hf_chat_template: true`. `ChatDataset` derives assistant masking from the tokenizer's chat template. VLM collators derive assistant labels from the processor template and do not accept `start_of_response_token`.
```

Reviewer: Please verify. These lines appear to be duplicated below.

Author: https://github.com/zeryx/Automodel/commit/010183a527466a9432592472dcba2da58053dff3 seems like this commit is on the blame for the duplicate content here. I can simply remove the second ChatDataset doc, but it seems that the notes section are not entirely duplicated.

Reviewer: It looks like the revised content is R91 to R118 with the addition of  R113 to R116 in the Note admonition.

Reviewer: ```suggestion
- Streams contiguous `seq_len` slices and supports optional BOS alignment and `.bos.idx` sidecar files
```

> - **RDR items**: `nemo_automodel.components.datasets.vlm.datasets.make_rdr_dataset` (HF: `quintend/rdr-items`)
> - **CORD-V2 receipts (Consolidated Receipt Dataset for Post-OCR Parsing)**: `nemo_automodel.components.datasets.vlm.datasets.make_cord_v2_dataset` (HF: `naver-clova-ix/cord-v2`)

Reviewer: ```suggestion
Built-in dataset makers return a Python list or a Hugging Face `Dataset`. Each yielded sample includes a `conversation` field:

- **RDR items**: `nemo_automodel.components.datasets.vlm.datasets.make_rdr_dataset` (Hugging Face: `quintend/rdr-items`)
- **CORD-V2 receipts (Consolidated Receipt Dataset for Post-OCR Parsing)**: `nemo_automodel.components.datasets.vlm.datasets.make_cord_v2_dataset` (Hugging Face: `naver-clova-ix/cord-v2`)
```

- **MedPix-VQA (Medical Pixel Question Answering)**: `nemo_automodel.components.datasets.vlm.datasets.make_medpix_dataset`

> - **Shopify product catalogue (product image to taxonomy category)**: `nemo_automodel.components.datasets.vlm.datasets.make_shopify_product_catalogue_dataset` (HF: `Shopify/product-catalogue`)

Reviewer: ```suggestion
- **Shopify product catalogue (product image to taxonomy category)**: `nemo_automodel.components.datasets.vlm.datasets.make_shopify_product_catalogue_dataset` (Hugging Face: `Shopify/product-catalogue`)
```

- **CommonVoice 17 (CV17) (audio)**: `nemo_automodel.components.datasets.audio.datasets.make_cv17_dataset`

> RDR and CORD-V2 return lists. MedPix, Shopify product catalogue, and CommonVoice 17 return Hugging Face `Dataset` objects with lazy transforms. Each sample follows the conversation schema expected by `apply_chat_template`, e.g.:

Reviewer: ```suggestion
RDR and CORD-V2 return lists. MedPix, Shopify product catalogue, and CommonVoice 17 return Hugging Face `Dataset` objects with lazy transforms. Each sample follows the conversation schema expected by `apply_chat_template`, for example:

```
